// roy/src/parser.js
/* Jison generated parser */
var parser = (function(){
undefined
var parser = {trace: function trace() { },
yy: {},
symbols_: {"error":2,"program":3,"EOF":4,"body":5,"line":6,"TERMINATOR":7,"statement":8,"expression":9,"COMMENT":10,"block":11,"INDENT":12,"OUTDENT":13,"doBody":14,"doLine":15,"BIND":16,"IDENTIFIER":17,"=":18,"RETURN":19,"doBlock":20,"letFunction":21,"letBinding":22,"dataDecl":23,"typeDecl":24,"macro":25,"innerExpression":26,"FN":27,"paramList":28,"optType":29,"MATCH":30,"caseList":31,"DO":32,"call":33,"ifThenElse":34,"(":35,")":36,"&":37,"[|":38,"|]":39,"accessor":40,"MATH":41,"CONCAT":42,"+":43,"-":44,"BOOLOP":45,"COMPARE":46,"WITH":47,"literal":48,"CASE":49,"pattern":50,"innerPattern":51,"identifier":52,"patternIdentifiers":53,"IF":54,"THEN":55,"ELSE":56,"DATA":57,"optParamList":58,"dataList":59,"|":60,"TYPE":61,"type":62,"{":63,"optTypePairs":64,"}":65,"keywordOrIdentifier":66,":":67,",":68,"MACRO":69,"LET":70,"param":71,"argList":72,"tuple":73,"tupleList":74,"NUMBER":75,"STRING":76,"BOOLEAN":77,"[":78,"optValues":79,"]":80,"optPairs":81,"arrayValues":82,"keyPairs":83,"optTerm":84,".":85,"$accept":0,"$end":1},
terminals_: {2:"error",4:"EOF",7:"TERMINATOR",10:"COMMENT",12:"INDENT",13:"OUTDENT",16:"BIND",17:"IDENTIFIER",18:"=",19:"RETURN",27:"FN",30:"MATCH",32:"DO",35:"(",36:")",37:"&",38:"[|",39:"|]",41:"MATH",42:"CONCAT",43:"+",44:"-",45:"BOOLOP",46:"COMPARE",47:"WITH",49:"CASE",54:"IF",55:"THEN",56:"ELSE",57:"DATA",60:"|",61:"TYPE",63:"{",65:"}",67:":",68:",",69:"MACRO",70:"LET",75:"NUMBER",76:"STRING",77:"BOOLEAN",78:"[",80:"]",85:"."},
productions_: [0,[3,1],[3,2],[5,1],[5,3],[5,2],[6,1],[6,1],[6,1],[11,3],[11,2],[14,1],[14,3],[14,2],[15,1],[15,4],[15,2],[20,3],[8,1],[8,1],[8,1],[8,1],[8,1],[9,1],[9,5],[9,5],[9,5],[9,3],[9,1],[26,1],[26,3],[26,4],[26,3],[26,1],[26,3],[26,3],[26,3],[26,3],[26,3],[26,3],[26,3],[26,1],[31,4],[31,6],[50,1],[50,1],[51,4],[53,1],[53,1],[53,2],[53,2],[34,7],[23,5],[23,7],[59,2],[59,4],[24,4],[62,1],[62,3],[64,0],[64,3],[64,5],[58,0],[58,1],[25,4],[25,4],[21,6],[21,6],[22,5],[22,7],[28,2],[28,1],[28,3],[28,2],[71,1],[71,5],[29,0],[29,2],[33,2],[33,4],[72,2],[72,1],[72,3],[72,2],[73,5],[74,1],[74,3],[48,1],[48,1],[48,1],[48,1],[48,3],[48,3],[79,0],[79,1],[82,1],[82,3],[81,0],[81,4],[81,1],[83,3],[83,5],[83,6],[84,0],[84,1],[40,1],[40,3],[40,5],[66,1],[66,1],[66,1],[52,1]],
performAction: function anonymous(yytext,yyleng,yylineno,yy,yystate,$$,_$) {

var $0 = $$.length - 1;
switch (yystate) {
case 1:return [];
break;
case 2:return $$[$0-1];
break;
case 3:this.$ = [$$[$0]];
break;
case 4:this.$ = $$[$0-2]; $$[$0-2].push($$[$0]);
break;
case 5:this.$ = $$[$0-1];
break;
case 6:this.$ = $$[$0];
break;
case 7:this.$ = $$[$0];
break;
case 8:this.$ = new yy.Comment($$[$0]);
break;
case 9:this.$ = $$[$0-1];
break;
case 10:this.$ = $$[$0-1];
break;
case 11:this.$ = [$$[$0]];
break;
case 12:this.$ = $$[$0-2]; $$[$0-2].push($$[$0]);
break;
case 13:this.$ = $$[$0-1];
break;
case 14:this.$ = $$[$0];
break;
case 15:this.$ = new yy.Bind($$[$0-2], $$[$0]);
break;
case 16:this.$ = new yy.Return($$[$0]);
break;
case 17:this.$ = $$[$0-1];
break;
case 18:this.$ = $$[$0];
break;
case 19:this.$ = $$[$0];
break;
case 20:this.$ = $$[$0];
break;
case 21:this.$ = $$[$0];
break;
case 22:this.$ = $$[$0];
break;
case 23:this.$ = $$[$0];
break;
case 24:this.$ = new yy.Function(undefined, $$[$0-3], [$$[$0]], $$[$0-2]);
break;
case 25:this.$ = new yy.Function(undefined, $$[$0-3], $$[$0], $$[$0-2]);
break;
case 26:this.$ = new yy.Match($$[$0-3], $$[$0-1]);
break;
case 27:this.$ = new yy.Do($$[$0-1], $$[$0]);
break;
case 28:this.$ = $$[$0];
break;
case 29:this.$ = $$[$0];
break;
case 30:this.$ = $$[$0-1];
break;
case 31:this.$ = new yy.Replacement($$[$0-1]);
break;
case 32:this.$ = new yy.Quoted($$[$0-1]);
break;
case 33:this.$ = $$[$0];
break;
case 34:this.$ = new yy.BinaryNumberOperator($$[$0-1], $$[$0-2], $$[$0]);
break;
case 35:this.$ = new yy.BinaryStringOperator($$[$0-1], $$[$0-2], $$[$0]);
break;
case 36:this.$ = new yy.BinaryNumberOperator($$[$0-1], $$[$0-2], $$[$0]);
break;
case 37:this.$ = new yy.BinaryNumberOperator($$[$0-1], $$[$0-2], $$[$0]);
break;
case 38:this.$ = new yy.BinaryBooleanOperator($$[$0-1], $$[$0-2], $$[$0]);
break;
case 39:this.$ = new yy.BinaryGenericOperator($$[$0-1], $$[$0-2], $$[$0]);
break;
case 40:this.$ = new yy.With($$[$0-2], $$[$0]);
break;
case 41:this.$ = $$[$0];
break;
case 42:this.$ = [new yy.Case($$[$0-2], $$[$0])];
break;
case 43:this.$ = $$[$0-5]; $$[$0-5].push(new yy.Case($$[$0-2], $$[$0]));
break;
case 44:this.$ = $$[$0];
break;
case 45:this.$ = new yy.Pattern($$[$0], []);
break;
case 46:this.$ = new yy.Pattern($$[$0-2], $$[$0-1]);
break;
case 47:this.$ = [$$[$0]];
break;
case 48:this.$ = [$$[$0]];
break;
case 49:this.$ = $$[$0-1]; $$[$0-1].push($$[$0]);
break;
case 50:this.$ = $$[$0-1]; $$[$0-1].push($$[$0]);
break;
case 51:this.$ = new yy.IfThenElse($$[$0-5], $$[$0-3], $$[$0]);
break;
case 52:this.$ = new yy.Data($$[$0-3], $$[$0-2], $$[$0]);
break;
case 53:this.$ = new yy.Data($$[$0-5], $$[$0-4], $$[$0-1]);
break;
case 54:this.$ = [new yy.Tag($$[$0-1], $$[$0])];
break;
case 55:this.$ = $$[$0-3]; $$[$0-3].push(new yy.Tag($$[$0-1], $$[$0]));
break;
case 56:this.$ = new yy.Type($$[$0-2], $$[$0]);
break;
case 57:this.$ = new yy.TypeName($$[$0]);
break;
case 58:this.$ = new yy.TypeObject($$[$0-1]);
break;
case 59:this.$ = {}
break;
case 60:this.$ = {}; this.$[$$[$0-2]] = $$[$0];
break;
case 61:this.$ = $$[$0-4]; $$[$0-4][$$[$0-2]] = $$[$0];
break;
case 62:this.$ = [];
break;
case 63:this.$ = $$[$0];
break;
case 64:this.$ = new yy.Macro($$[$0-2], [$$[$0]]);
break;
case 65:this.$ = new yy.Macro($$[$0-2], $$[$0]);
break;
case 66:this.$ = new yy.Function($$[$0-4], $$[$0-3], $$[$0], $$[$0-2]);
break;
case 67:this.$ = new yy.Function($$[$0-4], $$[$0-3], [$$[$0]], $$[$0-2]);
break;
case 68:this.$ = new yy.Let($$[$0-3], $$[$0], $$[$0-2]);
break;
case 69:this.$ = new yy.Let($$[$0-5], $$[$0-1], $$[$0-4]);
break;
case 70:this.$ = [];
break;
case 71:this.$ = [$$[$0]];
break;
case 72:this.$ = $$[$0-2];
break;
case 73:this.$ = $$[$0-1]; $$[$0-1].push($$[$0]);
break;
case 74:this.$ = new yy.Arg($$[$0]);
break;
case 75:this.$ = new yy.Arg($$[$0-3], $$[$0-1]);
break;
case 76:
break;
case 77:this.$ = $$[$0]
break;
case 78:this.$ = new yy.Call($$[$0-1], $$[$0]);
break;
case 79:this.$ = new yy.Call($$[$0-2], $$[$0]);
break;
case 80:this.$ = [];
break;
case 81:this.$ = [$$[$0]];
break;
case 82:this.$ = $$[$0-2];
break;
case 83:this.$ = $$[$0-1]; $$[$0-1].push($$[$0]);
break;
case 84:$$[$0-1].unshift($$[$0-3]); this.$ = new yy.Tuple($$[$0-1]);
break;
case 85:this.$ = [$$[$0]];
break;
case 86:this.$ = $$[$0-2]; $$[$0-2].push($$[$0]);
break;
case 87:this.$ = new yy.Number($$[$0]);
break;
case 88:this.$ = new yy.String($$[$0]);
break;
case 89:this.$ = new yy.Boolean($$[$0]);
break;
case 90:this.$ = $$[$0];
break;
case 91:this.$ = new yy.Array($$[$0-1]);
break;
case 92:this.$ = new yy.Object($$[$0-1]);
break;
case 93:this.$ = [];
break;
case 94:this.$ = $$[$0];
break;
case 95:this.$ = [$$[$0]];
break;
case 96:this.$ = $$[$0-2]; $$[$0-2].push($$[$0]);
break;
case 97:this.$ = {};
break;
case 98:this.$ = $$[$0-2];
break;
case 99:this.$ = $$[$0];
break;
case 100:this.$ = {}; this.$[$$[$0-2]] = $$[$0];
break;
case 101:this.$ = $$[$0-4]; $$[$0-4][$$[$0-2]] = $$[$0];
break;
case 102:this.$ = $$[$0-5]; $$[$0-5][$$[$0-2]] = $$[$0];
break;
case 103:
break;
case 104:
break;
case 105:this.$ = new yy.Identifier($$[$0]);
break;
case 106:this.$ = new yy.Access($$[$0-2], $$[$0]);
break;
case 107:this.$ = new yy.Access($$[$0-3], $$[$0]);
break;
case 108:this.$ = $$[$0];
break;
case 109:this.$ = $$[$0];
break;
case 110:this.$ = $$[$0];
break;
case 111:this.$ = new yy.Identifier($$[$0]);
break;
}
},
table: [{3:1,4:[1,2],5:3,6:4,8:5,9:6,10:[1,7],17:[1,29],21:8,22:9,23:10,24:11,25:12,26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],57:[1,19],61:[1,20],63:[1,35],69:[1,21],70:[1,18],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{1:[3]},{1:[2,1]},{4:[1,36],7:[1,37]},{4:[2,3],7:[2,3],13:[2,3]},{4:[2,6],7:[2,6],13:[2,6]},{4:[2,7],7:[2,7],13:[2,7]},{4:[2,8],7:[2,8],13:[2,8]},{4:[2,18],7:[2,18],13:[2,18]},{4:[2,19],7:[2,19],13:[2,19]},{4:[2,20],7:[2,20],13:[2,20]},{4:[2,21],7:[2,21],13:[2,21]},{4:[2,22],7:[2,22],13:[2,22]},{4:[2,23],7:[2,23],13:[2,23],36:[2,23],39:[2,23],41:[1,38],42:[1,39],43:[1,40],44:[1,41],45:[1,42],46:[1,43],47:[1,44],65:[2,23],68:[2,23],80:[2,23]},{17:[1,48],28:45,35:[1,46],71:47},{17:[1,29],26:49,34:22,35:[1,50],37:[1,24],38:[1,25],40:51,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{17:[1,29],26:52,34:22,35:[1,50],37:[1,24],38:[1,25],40:51,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{4:[2,28],7:[2,28],13:[2,28],36:[2,28],39:[2,28],65:[2,28],68:[2,28],80:[2,28]},{17:[1,53]},{17:[1,54]},{17:[1,55]},{17:[1,56]},{4:[2,29],7:[2,29],12:[2,29],13:[2,29],17:[2,29],35:[2,29],36:[2,29],37:[2,29],38:[2,29],39:[2,29],41:[2,29],42:[2,29],43:[2,29],44:[2,29],45:[2,29],46:[2,29],47:[2,29],54:[2,29],55:[2,29],63:[2,29],65:[2,29],68:[2,29],75:[2,29],76:[2,29],77:[2,29],78:[2,29],80:[2,29]},{9:57,17:[1,29],26:58,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{35:[1,59]},{9:60,17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{4:[2,33],7:[2,33],13:[2,33],17:[1,29],26:64,34:22,35:[1,63],36:[2,33],37:[1,24],38:[1,25],39:[2,33],40:51,41:[2,33],42:[2,33],43:[2,33],44:[2,33],45:[2,33],46:[2,33],47:[2,33],48:27,54:[1,28],63:[1,35],65:[2,33],68:[2,33],72:61,73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34],80:[2,33],85:[1,62]},{4:[2,41],7:[2,41],12:[2,41],13:[2,41],17:[2,41],35:[2,41],36:[2,41],37:[2,41],38:[2,41],39:[2,41],41:[2,41],42:[2,41],43:[2,41],44:[2,41],45:[2,41],46:[2,41],47:[2,41],54:[2,41],55:[2,41],63:[2,41],65:[2,41],68:[2,41],75:[2,41],76:[2,41],77:[2,41],78:[2,41],80:[2,41]},{17:[1,29],26:65,34:22,35:[1,50],37:[1,24],38:[1,25],40:51,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{4:[2,105],7:[2,105],12:[2,105],13:[2,105],17:[2,105],35:[2,105],36:[2,105],37:[2,105],38:[2,105],39:[2,105],41:[2,105],42:[2,105],43:[2,105],44:[2,105],45:[2,105],46:[2,105],47:[2,105],54:[2,105],55:[2,105],63:[2,105],65:[2,105],68:[2,105],75:[2,105],76:[2,105],77:[2,105],78:[2,105],80:[2,105],85:[2,105]},{4:[2,87],7:[2,87],12:[2,87],13:[2,87],17:[2,87],35:[2,87],36:[2,87],37:[2,87],38:[2,87],39:[2,87],41:[2,87],42:[2,87],43:[2,87],44:[2,87],45:[2,87],46:[2,87],47:[2,87],54:[2,87],55:[2,87],63:[2,87],65:[2,87],68:[2,87],75:[2,87],76:[2,87],77:[2,87],78:[2,87],80:[2,87]},{4:[2,88],7:[2,88],12:[2,88],13:[2,88],17:[2,88],35:[2,88],36:[2,88],37:[2,88],38:[2,88],39:[2,88],41:[2,88],42:[2,88],43:[2,88],44:[2,88],45:[2,88],46:[2,88],47:[2,88],54:[2,88],55:[2,88],63:[2,88],65:[2,88],68:[2,88],75:[2,88],76:[2,88],77:[2,88],78:[2,88],80:[2,88]},{4:[2,89],7:[2,89],12:[2,89],13:[2,89],17:[2,89],35:[2,89],36:[2,89],37:[2,89],38:[2,89],39:[2,89],41:[2,89],42:[2,89],43:[2,89],44:[2,89],45:[2,89],46:[2,89],47:[2,89],54:[2,89],55:[2,89],63:[2,89],65:[2,89],68:[2,89],75:[2,89],76:[2,89],77:[2,89],78:[2,89],80:[2,89]},{4:[2,90],7:[2,90],12:[2,90],13:[2,90],17:[2,90],35:[2,90],36:[2,90],37:[2,90],38:[2,90],39:[2,90],41:[2,90],42:[2,90],43:[2,90],44:[2,90],45:[2,90],46:[2,90],47:[2,90],54:[2,90],55:[2,90],63:[2,90],65:[2,90],68:[2,90],75:[2,90],76:[2,90],77:[2,90],78:[2,90],80:[2,90]},{9:68,17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34],79:66,80:[2,93],82:67},{12:[1,70],16:[1,74],17:[1,75],19:[1,73],65:[2,97],66:72,81:69,83:71},{1:[2,2]},{4:[2,5],6:76,7:[2,5],8:5,9:6,10:[1,7],13:[2,5],17:[1,29],21:8,22:9,23:10,24:11,25:12,26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],57:[1,19],61:[1,20],63:[1,35],69:[1,21],70:[1,18],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{17:[1,29],26:77,34:22,35:[1,50],37:[1,24],38:[1,25],40:51,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{17:[1,29],26:78,34:22,35:[1,50],37:[1,24],38:[1,25],40:51,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{17:[1,29],26:79,34:22,35:[1,50],37:[1,24],38:[1,25],40:51,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{17:[1,29],26:80,34:22,35:[1,50],37:[1,24],38:[1,25],40:51,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{17:[1,29],26:81,34:22,35:[1,50],37:[1,24],38:[1,25],40:51,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{17:[1,29],26:82,34:22,35:[1,50],37:[1,24],38:[1,25],40:51,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{17:[1,29],26:83,34:22,35:[1,50],37:[1,24],38:[1,25],40:51,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{17:[1,48],18:[2,76],29:84,35:[1,85],67:[1,87],71:86},{17:[1,89],36:[1,88]},{4:[2,71],7:[2,71],13:[2,71],17:[2,71],18:[2,71],35:[2,71],60:[2,71],67:[2,71]},{4:[2,74],7:[2,74],13:[2,74],17:[2,74],18:[2,74],35:[2,74],60:[2,74],67:[2,74]},{12:[1,90],41:[1,38],42:[1,39],43:[1,40],44:[1,41],45:[1,42],46:[1,43],47:[1,44]},{9:91,17:[1,29],26:58,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{4:[2,33],7:[2,33],12:[2,33],13:[2,33],17:[2,33],35:[2,33],36:[2,33],37:[2,33],38:[2,33],39:[2,33],41:[2,33],42:[2,33],43:[2,33],44:[2,33],45:[2,33],46:[2,33],47:[2,33],54:[2,33],55:[2,33],63:[2,33],65:[2,33],68:[2,33],75:[2,33],76:[2,33],77:[2,33],78:[2,33],80:[2,33],85:[1,62]},{12:[1,93],20:92,41:[1,38],42:[1,39],43:[1,40],44:[1,41],45:[1,42],46:[1,43],47:[1,44]},{17:[1,48],18:[2,76],28:94,29:95,35:[1,46],67:[1,87],71:47},{17:[1,48],18:[2,62],28:97,35:[1,46],58:96,71:47},{18:[1,98]},{18:[1,99]},{36:[1,100]},{36:[2,23],41:[1,38],42:[1,39],43:[1,40],44:[1,41],45:[1,42],46:[1,43],47:[1,44],68:[1,101]},{9:102,17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{39:[1,103]},{4:[2,78],7:[2,78],13:[2,78],17:[1,29],26:105,34:22,35:[1,104],36:[2,78],37:[1,24],38:[1,25],39:[2,78],40:51,48:27,54:[1,28],63:[1,35],65:[2,78],68:[2,78],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34],80:[2,78]},{16:[1,74],17:[1,75],19:[1,73],66:106},{9:91,17:[1,29],26:58,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],36:[1,107],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{4:[2,81],7:[2,81],13:[2,81],17:[2,81],35:[2,81],36:[2,81],37:[2,81],38:[2,81],39:[2,81],41:[1,38],42:[1,39],43:[1,40],44:[1,41],45:[1,42],46:[1,43],47:[1,44],54:[2,81],63:[2,81],65:[2,81],68:[2,81],75:[2,81],76:[2,81],77:[2,81],78:[2,81],80:[2,81]},{41:[1,38],42:[1,39],43:[1,40],44:[1,41],45:[1,42],46:[1,43],47:[1,44],55:[1,108]},{80:[1,109]},{68:[1,110],80:[2,94]},{68:[2,95],80:[2,95]},{65:[1,111]},{16:[1,74],17:[1,75],19:[1,73],66:72,83:112},{7:[1,114],65:[2,99],68:[1,113]},{67:[1,115]},{4:[2,108],7:[2,108],12:[2,108],13:[2,108],17:[2,108],35:[2,108],36:[2,108],37:[2,108],38:[2,108],39:[2,108],41:[2,108],42:[2,108],43:[2,108],44:[2,108],45:[2,108],46:[2,108],47:[2,108],54:[2,108],55:[2,108],63:[2,108],65:[2,108],67:[2,108],68:[2,108],75:[2,108],76:[2,108],77:[2,108],78:[2,108],80:[2,108],85:[2,108]},{4:[2,109],7:[2,109],12:[2,109],13:[2,109],17:[2,109],35:[2,109],36:[2,109],37:[2,109],38:[2,109],39:[2,109],41:[2,109],42:[2,109],43:[2,109],44:[2,109],45:[2,109],46:[2,109],47:[2,109],54:[2,109],55:[2,109],63:[2,109],65:[2,109],67:[2,109],68:[2,109],75:[2,109],76:[2,109],77:[2,109],78:[2,109],80:[2,109],85:[2,109]},{4:[2,110],7:[2,110],12:[2,110],13:[2,110],17:[2,110],35:[2,110],36:[2,110],37:[2,110],38:[2,110],39:[2,110],41:[2,110],42:[2,110],43:[2,110],44:[2,110],45:[2,110],46:[2,110],47:[2,110],54:[2,110],55:[2,110],63:[2,110],65:[2,110],67:[2,110],68:[2,110],75:[2,110],76:[2,110],77:[2,110],78:[2,110],80:[2,110],85:[2,110]},{4:[2,4],7:[2,4],13:[2,4]},{4:[2,34],7:[2,34],12:[2,34],13:[2,34],17:[2,34],35:[2,34],36:[2,34],37:[2,34],38:[2,34],39:[2,34],41:[2,34],42:[2,34],43:[2,34],44:[2,34],45:[2,34],46:[2,34],47:[2,34],54:[2,34],55:[2,34],63:[2,34],65:[2,34],68:[2,34],75:[2,34],76:[2,34],77:[2,34],78:[2,34],80:[2,34]},{4:[2,35],7:[2,35],12:[2,35],13:[2,35],17:[2,35],35:[2,35],36:[2,35],37:[2,35],38:[2,35],39:[2,35],41:[2,35],42:[2,35],43:[2,35],44:[2,35],45:[2,35],46:[2,35],47:[2,35],54:[2,35],55:[2,35],63:[2,35],65:[2,35],68:[2,35],75:[2,35],76:[2,35],77:[2,35],78:[2,35],80:[2,35]},{4:[2,36],7:[2,36],12:[2,36],13:[2,36],17:[2,36],35:[2,36],36:[2,36],37:[2,36],38:[2,36],39:[2,36],41:[1,38],42:[1,39],43:[2,36],44:[2,36],45:[2,36],46:[2,36],47:[2,36],54:[2,36],55:[2,36],63:[2,36],65:[2,36],68:[2,36],75:[2,36],76:[2,36],77:[2,36],78:[2,36],80:[2,36]},{4:[2,37],7:[2,37],12:[2,37],13:[2,37],17:[2,37],35:[2,37],36:[2,37],37:[2,37],38:[2,37],39:[2,37],41:[1,38],42:[1,39],43:[2,37],44:[2,37],45:[2,37],46:[2,37],47:[2,37],54:[2,37],55:[2,37],63:[2,37],65:[2,37],68:[2,37],75:[2,37],76:[2,37],77:[2,37],78:[2,37],80:[2,37]},{4:[2,38],7:[2,38],12:[2,38],13:[2,38],17:[2,38],35:[2,38],36:[2,38],37:[2,38],38:[2,38],39:[2,38],41:[1,38],42:[1,39],43:[1,40],44:[1,41],45:[2,38],46:[1,43],47:[1,44],54:[2,38],55:[2,38],63:[2,38],65:[2,38],68:[2,38],75:[2,38],76:[2,38],77:[2,38],78:[2,38],80:[2,38]},{4:[2,39],7:[2,39],12:[2,39],13:[2,39],17:[2,39],35:[2,39],36:[2,39],37:[2,39],38:[2,39],39:[2,39],41:[1,38],42:[1,39],43:[1,40],44:[1,41],45:[2,39],46:[2,39],47:[2,39],54:[2,39],55:[2,39],63:[2,39],65:[2,39],68:[2,39],75:[2,39],76:[2,39],77:[2,39],78:[2,39],80:[2,39]},{4:[2,40],7:[2,40],12:[2,40],13:[2,40],17:[2,40],35:[2,40],36:[2,40],37:[2,40],38:[2,40],39:[2,40],41:[1,38],42:[1,39],43:[1,40],44:[1,41],45:[2,40],46:[2,40],47:[2,40],54:[2,40],55:[2,40],63:[2,40],65:[2,40],68:[2,40],75:[2,40],76:[2,40],77:[2,40],78:[2,40],80:[2,40]},{18:[1,116]},{17:[1,89],36:[1,117]},{4:[2,73],7:[2,73],13:[2,73],17:[2,73],18:[2,73],35:[2,73],60:[2,73],67:[2,73]},{17:[1,119],62:118,63:[1,120]},{4:[2,70],7:[2,70],13:[2,70],17:[2,70],18:[2,70],35:[2,70],60:[2,70],67:[2,70]},{67:[1,121]},{31:122,49:[1,123]},{36:[1,124]},{4:[2,27],7:[2,27],13:[2,27],36:[2,27],39:[2,27],65:[2,27],68:[2,27],80:[2,27]},{6:127,8:5,9:6,10:[1,7],14:125,15:126,16:[1,128],17:[1,29],19:[1,129],21:8,22:9,23:10,24:11,25:12,26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],57:[1,19],61:[1,20],63:[1,35],69:[1,21],70:[1,18],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{17:[1,48],18:[2,76],29:130,35:[1,85],67:[1,87],71:86},{18:[1,131]},{18:[1,132]},{4:[2,63],7:[2,63],13:[2,63],17:[1,48],18:[2,63],35:[1,85],60:[2,63],71:86},{17:[1,119],62:133,63:[1,120]},{9:134,11:135,12:[1,136],17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{4:[2,30],7:[2,30],13:[2,30],17:[1,29],26:64,34:22,35:[1,63],36:[2,30],37:[1,24],38:[1,25],39:[2,30],40:51,41:[2,30],42:[2,30],43:[2,30],44:[2,30],45:[2,30],46:[2,30],47:[2,30],48:27,54:[1,28],63:[1,35],65:[2,30],68:[2,30],72:137,73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34],80:[2,30],85:[1,138]},{17:[1,29],26:140,34:22,35:[1,50],37:[1,24],38:[1,25],40:51,48:27,54:[1,28],63:[1,35],73:33,74:139,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{36:[1,141]},{4:[2,32],7:[2,32],12:[2,32],13:[2,32],17:[2,32],35:[2,32],36:[2,32],37:[2,32],38:[2,32],39:[2,32],41:[2,32],42:[2,32],43:[2,32],44:[2,32],45:[2,32],46:[2,32],47:[2,32],54:[2,32],55:[2,32],63:[2,32],65:[2,32],68:[2,32],75:[2,32],76:[2,32],77:[2,32],78:[2,32],80:[2,32]},{9:91,17:[1,29],26:58,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],36:[1,142],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{4:[2,83],7:[2,83],13:[2,83],17:[2,83],35:[2,83],36:[2,83],37:[2,83],38:[2,83],39:[2,83],41:[1,38],42:[1,39],43:[1,40],44:[1,41],45:[1,42],46:[1,43],47:[1,44],54:[2,83],63:[2,83],65:[2,83],68:[2,83],75:[2,83],76:[2,83],77:[2,83],78:[2,83],80:[2,83]},{4:[2,106],7:[2,106],12:[2,106],13:[2,106],17:[2,106],35:[2,106],36:[2,106],37:[2,106],38:[2,106],39:[2,106],41:[2,106],42:[2,106],43:[2,106],44:[2,106],45:[2,106],46:[2,106],47:[2,106],54:[2,106],55:[2,106],63:[2,106],65:[2,106],68:[2,106],75:[2,106],76:[2,106],77:[2,106],78:[2,106],80:[2,106],85:[2,106]},{4:[2,80],7:[2,80],13:[2,80],17:[2,80],35:[2,80],36:[2,80],37:[2,80],38:[2,80],39:[2,80],54:[2,80],63:[2,80],65:[2,80],68:[2,80],75:[2,80],76:[2,80],77:[2,80],78:[2,80],80:[2,80]},{11:143,12:[1,136]},{4:[2,91],7:[2,91],12:[2,91],13:[2,91],17:[2,91],35:[2,91],36:[2,91],37:[2,91],38:[2,91],39:[2,91],41:[2,91],42:[2,91],43:[2,91],44:[2,91],45:[2,91],46:[2,91],47:[2,91],54:[2,91],55:[2,91],63:[2,91],65:[2,91],68:[2,91],75:[2,91],76:[2,91],77:[2,91],78:[2,91],80:[2,91]},{9:144,17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{4:[2,92],7:[2,92],12:[2,92],13:[2,92],17:[2,92],35:[2,92],36:[2,92],37:[2,92],38:[2,92],39:[2,92],41:[2,92],42:[2,92],43:[2,92],44:[2,92],45:[2,92],46:[2,92],47:[2,92],54:[2,92],55:[2,92],63:[2,92],65:[2,92],68:[2,92],75:[2,92],76:[2,92],77:[2,92],78:[2,92],80:[2,92]},{7:[1,114],13:[1,145],68:[1,113]},{16:[1,74],17:[1,75],19:[1,73],66:146},{7:[1,148],16:[2,103],17:[2,103],19:[2,103],84:147},{9:149,17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{9:150,11:151,12:[1,136],17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{4:[2,72],7:[2,72],13:[2,72],17:[2,72],18:[2,72],35:[2,72],60:[2,72],67:[2,72]},{18:[2,77]},{4:[2,57],7:[2,57],13:[2,57],18:[2,57],36:[2,57],65:[2,57],68:[2,57]},{16:[1,74],17:[1,75],19:[1,73],64:152,65:[2,59],66:153,68:[2,59]},{17:[1,119],62:154,63:[1,120]},{7:[1,156],13:[1,155]},{17:[1,161],35:[1,160],50:157,51:158,52:159},{4:[2,30],7:[2,30],12:[2,30],13:[2,30],17:[2,30],35:[2,30],36:[2,30],37:[2,30],38:[2,30],39:[2,30],41:[2,30],42:[2,30],43:[2,30],44:[2,30],45:[2,30],46:[2,30],47:[2,30],54:[2,30],55:[2,30],63:[2,30],65:[2,30],68:[2,30],75:[2,30],76:[2,30],77:[2,30],78:[2,30],80:[2,30],85:[1,138]},{7:[1,163],13:[1,162]},{7:[2,11],13:[2,11]},{7:[2,14],13:[2,14]},{17:[1,164]},{9:165,17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{18:[1,166]},{9:167,12:[1,168],17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{12:[1,170],17:[1,171],59:169},{4:[2,56],7:[2,56],13:[2,56]},{4:[2,64],7:[2,64],13:[2,64]},{4:[2,65],7:[2,65],13:[2,65]},{5:172,6:4,8:5,9:6,10:[1,7],13:[1,173],17:[1,29],21:8,22:9,23:10,24:11,25:12,26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],57:[1,19],61:[1,20],63:[1,35],69:[1,21],70:[1,18],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{4:[2,79],7:[2,79],13:[2,79],17:[1,29],26:105,34:22,35:[1,104],36:[2,79],37:[1,24],38:[1,25],39:[2,79],40:51,48:27,54:[1,28],63:[1,35],65:[2,79],68:[2,79],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34],80:[2,79]},{16:[1,74],17:[1,75],19:[1,73],66:174},{36:[1,175],68:[1,176]},{36:[2,85],41:[1,38],42:[1,39],43:[1,40],44:[1,41],45:[1,42],46:[1,43],47:[1,44],68:[2,85]},{4:[2,31],7:[2,31],12:[2,31],13:[2,31],17:[2,31],35:[2,31],36:[2,31],37:[2,31],38:[2,31],39:[2,31],41:[2,31],42:[2,31],43:[2,31],44:[2,31],45:[2,31],46:[2,31],47:[2,31],54:[2,31],55:[2,31],63:[2,31],65:[2,31],68:[2,31],75:[2,31],76:[2,31],77:[2,31],78:[2,31],80:[2,31]},{4:[2,82],7:[2,82],13:[2,82],17:[2,82],35:[2,82],36:[2,82],37:[2,82],38:[2,82],39:[2,82],54:[2,82],63:[2,82],65:[2,82],68:[2,82],75:[2,82],76:[2,82],77:[2,82],78:[2,82],80:[2,82]},{7:[1,177]},{68:[2,96],80:[2,96]},{7:[1,178]},{67:[1,179]},{16:[1,74],17:[1,75],19:[1,73],66:180},{16:[2,104],17:[2,104],19:[2,104]},{7:[2,100],13:[2,100],65:[2,100],68:[2,100]},{4:[2,24],7:[2,24],13:[2,24],36:[2,24],39:[2,24],65:[2,24],68:[2,24],80:[2,24]},{4:[2,25],7:[2,25],13:[2,25],36:[2,25],39:[2,25],65:[2,25],68:[2,25],80:[2,25]},{65:[1,181],68:[1,182]},{67:[1,183]},{36:[1,184]},{4:[2,26],7:[2,26],13:[2,26],36:[2,26],39:[2,26],65:[2,26],68:[2,26],80:[2,26]},{49:[1,185]},{18:[1,186]},{18:[2,44]},{18:[2,45]},{17:[1,161],52:187},{17:[2,111],18:[2,111],35:[2,111],36:[2,111]},{4:[2,17],7:[2,17],13:[2,17],36:[2,17],39:[2,17],65:[2,17],68:[2,17],80:[2,17]},{6:127,7:[2,13],8:5,9:6,10:[1,7],13:[2,13],15:188,16:[1,128],17:[1,29],19:[1,129],21:8,22:9,23:10,24:11,25:12,26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],57:[1,19],61:[1,20],63:[1,35],69:[1,21],70:[1,18],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{18:[1,189]},{7:[2,16],13:[2,16]},{9:191,11:190,12:[1,136],17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{4:[2,68],7:[2,68],13:[2,68]},{9:192,17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{4:[2,52],7:[2,52],13:[2,52],60:[1,193]},{17:[1,171],59:194},{4:[2,62],7:[2,62],13:[2,62],17:[1,48],28:97,35:[1,46],58:195,60:[2,62],71:47},{7:[1,37],13:[1,196]},{4:[2,10],7:[2,10],12:[2,10],13:[2,10],17:[2,10],35:[2,10],36:[2,10],37:[2,10],38:[2,10],39:[2,10],41:[2,10],42:[2,10],43:[2,10],44:[2,10],45:[2,10],46:[2,10],47:[2,10],54:[2,10],55:[2,10],63:[2,10],65:[2,10],68:[2,10],75:[2,10],76:[2,10],77:[2,10],78:[2,10],80:[2,10]},{4:[2,107],7:[2,107],12:[2,107],13:[2,107],17:[2,107],35:[2,107],36:[2,107],37:[2,107],38:[2,107],39:[2,107],41:[2,107],42:[2,107],43:[2,107],44:[2,107],45:[2,107],46:[2,107],47:[2,107],54:[2,107],55:[2,107],63:[2,107],65:[2,107],68:[2,107],75:[2,107],76:[2,107],77:[2,107],78:[2,107],80:[2,107],85:[2,107]},{4:[2,84],7:[2,84],12:[2,84],13:[2,84],17:[2,84],35:[2,84],36:[2,84],37:[2,84],38:[2,84],39:[2,84],41:[2,84],42:[2,84],43:[2,84],44:[2,84],45:[2,84],46:[2,84],47:[2,84],54:[2,84],55:[2,84],63:[2,84],65:[2,84],68:[2,84],75:[2,84],76:[2,84],77:[2,84],78:[2,84],80:[2,84]},{17:[1,29],26:197,34:22,35:[1,50],37:[1,24],38:[1,25],40:51,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{56:[1,198]},{65:[2,98]},{9:199,17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{67:[1,200]},{4:[2,58],7:[2,58],13:[2,58],18:[2,58],36:[2,58],65:[2,58],68:[2,58]},{16:[1,74],17:[1,75],19:[1,73],66:201},{17:[1,119],62:202,63:[1,120]},{4:[2,75],7:[2,75],13:[2,75],17:[2,75],18:[2,75],35:[2,75],60:[2,75],67:[2,75]},{17:[1,161],35:[1,160],50:203,51:158,52:159},{9:204,17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{17:[1,161],35:[1,160],51:207,52:206,53:205},{7:[2,12],13:[2,12]},{9:208,17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{4:[2,66],7:[2,66],13:[2,66]},{4:[2,67],7:[2,67],13:[2,67]},{13:[1,209]},{17:[1,210]},{13:[1,211],60:[1,193]},{4:[2,54],7:[2,54],13:[2,54],60:[2,54]},{4:[2,9],7:[2,9],12:[2,9],13:[2,9],17:[2,9],35:[2,9],36:[2,9],37:[2,9],38:[2,9],39:[2,9],41:[2,9],42:[2,9],43:[2,9],44:[2,9],45:[2,9],46:[2,9],47:[2,9],54:[2,9],55:[2,9],63:[2,9],65:[2,9],68:[2,9],75:[2,9],76:[2,9],77:[2,9],78:[2,9],80:[2,9]},{36:[2,86],41:[1,38],42:[1,39],43:[1,40],44:[1,41],45:[1,42],46:[1,43],47:[1,44],68:[2,86]},{11:212,12:[1,136]},{7:[2,101],13:[2,101],65:[2,101],68:[2,101]},{9:213,17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{67:[1,214]},{65:[2,60],68:[2,60]},{18:[1,215]},{7:[2,42],13:[2,42]},{17:[1,161],35:[1,160],36:[1,216],51:217,52:218},{17:[2,47],35:[2,47],36:[2,47]},{17:[2,48],35:[2,48],36:[2,48]},{7:[2,15],13:[2,15]},{4:[2,69],7:[2,69],13:[2,69]},{4:[2,62],7:[2,62],13:[2,62],17:[1,48],28:97,35:[1,46],58:219,60:[2,62],71:47},{4:[2,53],7:[2,53],13:[2,53]},{4:[2,51],7:[2,51],12:[2,51],13:[2,51],17:[2,51],35:[2,51],36:[2,51],37:[2,51],38:[2,51],39:[2,51],41:[2,51],42:[2,51],43:[2,51],44:[2,51],45:[2,51],46:[2,51],47:[2,51],54:[2,51],55:[2,51],63:[2,51],65:[2,51],68:[2,51],75:[2,51],76:[2,51],77:[2,51],78:[2,51],80:[2,51]},{7:[2,102],13:[2,102],65:[2,102],68:[2,102]},{17:[1,119],62:220,63:[1,120]},{9:221,17:[1,29],26:13,27:[1,14],30:[1,15],32:[1,16],33:17,34:22,35:[1,23],37:[1,24],38:[1,25],40:26,48:27,54:[1,28],63:[1,35],73:33,75:[1,30],76:[1,31],77:[1,32],78:[1,34]},{17:[2,46],18:[2,46],35:[2,46],36:[2,46]},{17:[2,49],35:[2,49],36:[2,49]},{17:[2,50],35:[2,50],36:[2,50]},{4:[2,55],7:[2,55],13:[2,55],60:[2,55]},{65:[2,61],68:[2,61]},{7:[2,43],13:[2,43]}],
defaultActions: {2:[2,1],36:[2,2],118:[2,77],158:[2,44],159:[2,45],178:[2,98]},
parseError: function parseError(str, hash) {
    throw new Error(str);
},
parse: function parse(input) {
    var self = this,
        stack = [0],
        vstack = [null], // semantic value stack
        lstack = [], // location stack
        table = this.table,
        yytext = '',
        yylineno = 0,
        yyleng = 0,
        recovering = 0,
        TERROR = 2,
        EOF = 1;

    //this.reductionCount = this.shiftCount = 0;

    this.lexer.setInput(input);
    this.lexer.yy = this.yy;
    this.yy.lexer = this.lexer;
    if (typeof this.lexer.yylloc == 'undefined')
        this.lexer.yylloc = {};
    var yyloc = this.lexer.yylloc;
    lstack.push(yyloc);

    if (typeof this.yy.parseError === 'function')
        this.parseError = this.yy.parseError;

    function popStack (n) {
        stack.length = stack.length - 2*n;
        vstack.length = vstack.length - n;
        lstack.length = lstack.length - n;
    }

    function lex() {
        var token;
        token = self.lexer.lex() || 1; // $end = 1
        // if token isn't its numeric value, convert
        if (typeof token !== 'number') {
            token = self.symbols_[token] || token;
        }
        return token;
    }

    var symbol, preErrorSymbol, state, action, a, r, yyval={},p,len,newState, expected;
    while (true) {
        // retreive state number from top of stack
        state = stack[stack.length-1];

        // use default actions if available
        if (this.defaultActions[state]) {
            action = this.defaultActions[state];
        } else {
            if (symbol == null)
                symbol = lex();
            // read action for current state and first input
            action = table[state] && table[state][symbol];
        }

        // handle parse error
        _handle_error:
        if (typeof action === 'undefined' || !action.length || !action[0]) {

            if (!recovering) {
                // Report error
                expected = [];
                for (p in table[state]) if (this.terminals_[p] && p > 2) {
                    expected.push("'"+this.terminals_[p]+"'");
                }
                var errStr = '';
                if (this.lexer.showPosition) {
                    errStr = 'Parse error on line '+(yylineno+1)+":\n"+this.lexer.showPosition()+"\nExpecting "+expected.join(', ') + ", got '" + this.terminals_[symbol]+ "'";
                } else {
                    errStr = 'Parse error on line '+(yylineno+1)+": Unexpected " +
                                  (symbol == 1 /*EOF*/ ? "end of input" :
                                              ("'"+(this.terminals_[symbol] || symbol)+"'"));
                }
                this.parseError(errStr,
                    {text: this.lexer.match, token: this.terminals_[symbol] || symbol, line: this.lexer.yylineno, loc: yyloc, expected: expected});
            }

            // just recovered from another error
            if (recovering == 3) {
                if (symbol == EOF) {
                    throw new Error(errStr || 'Parsing halted.');
                }

                // discard current lookahead and grab another
                yyleng = this.lexer.yyleng;
                yytext = this.lexer.yytext;
                yylineno = this.lexer.yylineno;
                yyloc = this.lexer.yylloc;
                symbol = lex();
            }

            // try to recover from error
            while (1) {
                // check for error recovery rule in this state
                if ((TERROR.toString()) in table[state]) {
                    break;
                }
                if (state == 0) {
                    throw new Error(errStr || 'Parsing halted.');
                }
                popStack(1);
                state = stack[stack.length-1];
            }

            preErrorSymbol = symbol; // save the lookahead token
            symbol = TERROR;         // insert generic error symbol as new lookahead
            state = stack[stack.length-1];
            action = table[state] && table[state][TERROR];
            recovering = 3; // allow 3 real symbols to be shifted before reporting a new error
        }

        // this shouldn't happen, unless resolve defaults are off
        if (action[0] instanceof Array && action.length > 1) {
            throw new Error('Parse Error: multiple actions possible at state: '+state+', token: '+symbol);
        }

        switch (action[0]) {

            case 1: // shift
                //this.shiftCount++;

                stack.push(symbol);
                vstack.push(this.lexer.yytext);
                lstack.push(this.lexer.yylloc);
                stack.push(action[1]); // push state
                symbol = null;
                if (!preErrorSymbol) { // normal execution/no error
                    yyleng = this.lexer.yyleng;
                    yytext = this.lexer.yytext;
                    yylineno = this.lexer.yylineno;
                    yyloc = this.lexer.yylloc;
                    if (recovering > 0)
                        recovering--;
                } else { // error just occurred, resume old lookahead f/ before error
                    symbol = preErrorSymbol;
                    preErrorSymbol = null;
                }
                break;

            case 2: // reduce
                //this.reductionCount++;

                len = this.productions_[action[1]][1];

                // perform semantic action
                yyval.$ = vstack[vstack.length-len]; // default to $$ = $1
                // default location, uses first token for firsts, last for lasts
                yyval._$ = {
                    first_line: lstack[lstack.length-(len||1)].first_line,
                    last_line: lstack[lstack.length-1].last_line,
                    first_column: lstack[lstack.length-(len||1)].first_column,
                    last_column: lstack[lstack.length-1].last_column
                };
                r = this.performAction.call(yyval, yytext, yyleng, yylineno, this.yy, action[1], vstack, lstack);

                if (typeof r !== 'undefined') {
                    return r;
                }

                // pop off stack
                if (len) {
                    stack = stack.slice(0,-1*len*2);
                    vstack = vstack.slice(0, -1*len);
                    lstack = lstack.slice(0, -1*len);
                }

                stack.push(this.productions_[action[1]][0]);    // push nonterminal (reduce)
                vstack.push(yyval.$);
                lstack.push(yyval._$);
                // goto new state = table[STATE][NONTERMINAL]
                newState = table[stack[stack.length-2]][stack[stack.length-1]];
                stack.push(newState);
                break;

            case 3: // accept
                return true;
        }

    }

    return true;
}};
return parser;
})();
if (typeof require !== 'undefined' && typeof exports !== 'undefined') {
exports.parser = parser;
exports.parse = function () { return parser.parse.apply(parser, arguments); }
exports.main = function commonjsMain(args) {
    if (!args[1])
        throw new Error('Usage: '+args[0]+' FILE');
    if (typeof process !== 'undefined') {
        var source = require('fs').readFileSync(require('path').join(process.cwd(), args[1]), "utf8");
    } else {
        var cwd = require("file").path(require("file").cwd());
        var source = cwd.join(args[1]).read({charset: "utf-8"});
    }
    return exports.parser.parse(source);
}
if (typeof module !== 'undefined' && require.main === module) {
  exports.main(typeof process !== 'undefined' ? process.argv.slice(1) : require("system").args);
}
}
